File: day04/day04.rs
use std::fs;

const DIRECTIONS: [(i32, i32); 8] = [
    (-1, 0),  // N
    (-1, 1),  // NE
    (0, 1),   // E
    (1, 1),   // SE
    (1, 0),   // S
    (1, -1),  // SW
    (0, -1),  // W
    (-1, -1), // NW
];

fn parse_grid(input: &str) -> Vec<Vec<char>> {
    input
        .trim()
        .lines()
        .map(|line| line.chars().collect())
        .collect()
}

fn count_adjacent_rolls(grid: &[Vec<char>], row: usize, col: usize) -> usize {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;

    DIRECTIONS
        .iter()
        .filter(|(dr, dc)| {
            let new_row = row as i32 + dr;
            let new_col = col as i32 + dc;

            new_row >= 0
                && new_row < rows
                && new_col >= 0
                && new_col < cols
                && grid[new_row as usize][new_col as usize] == '@'
        })
        .count()
}

fn find_accessible_rolls(grid: &[Vec<char>]) -> Vec<(usize, usize)> {
    let rows = grid.len();
    let cols = grid[0].len();

    (0..rows)
        .flat_map(|row| (0..cols).map(move |col| (row, col)))
        .filter(|&(row, col)| grid[row][col] == '@' && count_adjacent_rolls(grid, row, col) < 4)
        .collect()
}

fn solve_part1(grid: &[Vec<char>]) -> usize {
    find_accessible_rolls(grid).len()
}

fn solve_part2(grid: &[Vec<char>]) -> usize {
    let mut grid = grid.to_vec();
    let mut total_removed = 0;

    loop {
        let accessible = find_accessible_rolls(&grid);

        if accessible.is_empty() {
            break;
        }

        for (row, col) in &accessible {
            grid[*row][*col] = '.';
        }

        total_removed += accessible.len();
    }

    total_removed
}

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input.txt");
    let grid = parse_grid(&input);

    println!("Part 1: {}", solve_part1(&grid));
    println!("Part 2: {}", solve_part2(&grid));
}

File: day04/Day04.cs
var input = File.ReadAllText("input.txt");
var grid = input.Trim().Split('\n').Select(line => line.ToCharArray()).ToArray();

int[] dr = [-1, -1, 0, 1, 1, 1, 0, -1];
int[] dc = [0, 1, 1, 1, 0, -1, -1, -1];

int CountAdjacentRolls(char[][] g, int row, int col)
{
    var rows = g.Length;
    var cols = g[0].Length;
    var count = 0;

    for (var i = 0; i < 8; i++)
    {
        var newRow = row + dr[i];
        var newCol = col + dc[i];

        if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && g[newRow][newCol] == '@')
            count++;
    }

    return count;
}

List<(int Row, int Col)> FindAccessibleRolls(char[][] g)
{
    var rows = g.Length;
    var cols = g[0].Length;
    var accessible = new List<(int, int)>();

    for (var row = 0; row < rows; row++)
    {
        for (var col = 0; col < cols; col++)
        {
            if (g[row][col] == '@' && CountAdjacentRolls(g, row, col) < 4)
                accessible.Add((row, col));
        }
    }

    return accessible;
}

// Part 1: Count accessible rolls
var part1 = FindAccessibleRolls(grid).Count;
Console.WriteLine($"Part 1: {part1}");

// Part 2: Iteratively remove accessible rolls until no more can be removed
var mutableGrid = grid.Select(row => (char[])row.Clone()).ToArray();
var totalRemoved = 0;

while (true)
{
    var accessible = FindAccessibleRolls(mutableGrid);
    if (accessible.Count == 0)
        break;

    foreach (var (row, col) in accessible)
        mutableGrid[row][col] = '.';

    totalRemoved += accessible.Count;
}

Console.WriteLine($"Part 2: {totalRemoved}");
